File: prototype/rails/src/test/models/placeholder_image_test.rb
require "test_helper"

class PlaceholderImageTest < ActiveSupport::TestCase
  test "same title renders the same svg" do
    assert_equal PlaceholderImage.svg("Blue Heron"), PlaceholderImage.svg("Blue Heron")
  end

  test "different titles render different svgs" do
    refute_equal PlaceholderImage.svg("Blue Heron"), PlaceholderImage.svg("Red Fox")
  end

  test "svg carries the title as an accessible label" do
    svg = PlaceholderImage.svg("Mug & Bowl")

    assert_includes svg, 'aria-label="Mug &amp; Bowl"'
    assert svg.lstrip.start_with?("<svg")
  end

  test "data uri is base64 svg" do
    uri = PlaceholderImage.data_uri("Blue Heron")
    prefix = "data:image/svg+xml;base64,"

    assert uri.start_with?(prefix)
    assert_equal PlaceholderImage.svg("Blue Heron"), Base64.strict_decode64(uri.delete_prefix(prefix))
  end
end

File: app/src/app/Domain/Listings/PlaceholderImageTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Listings;

it('renders the same svg for the same title', function (): void {
    expect(PlaceholderImage::svg('Blue Heron'))->toBe(PlaceholderImage::svg('Blue Heron'));
});

it('renders different svgs for different titles', function (): void {
    expect(PlaceholderImage::svg('Blue Heron'))->not->toBe(PlaceholderImage::svg('Red Fox'));
});

it('carries the title as an accessible label', function (): void {
    $svg = PlaceholderImage::svg('Mug & Bowl');

    expect($svg)->toContain('aria-label="Mug &amp; Bowl"')
        ->and($svg)->toStartWith('<svg');
});

it('encodes the data uri as base64 svg', function (): void {
    $uri = PlaceholderImage::dataUri('Blue Heron');

    expect($uri)->toStartWith('data:image/svg+xml;base64,')
        ->and(base64_decode(substr($uri, strlen('data:image/svg+xml;base64,')), true))
        ->toBe(PlaceholderImage::svg('Blue Heron'));
});
